#8 - convert resources into static api

plencka opened the pull request:
This PR should close #8.

Currently build script should convert existing .csv files to .json according to structure defined in DataBuilders.

Static output for current resources should result in:
```
+ public
  + api
    + faculties
       + faculties.json
       + wydzial-techniczny
           + companies.json
```

krzysztofrewak commented on backend/Application/Application.php:
```diff
@@ -0,0 +1,65 @@
+            $this->directoryManager->getResourcePath($this->facultyBuilder->getSource()->getRelativePath())
+            . $this->facultyBuilder->getSource()->getFileName()
+        );
+        $faculties = $this->facultyBuilder->buildFromData($facultyCsvData);
```
I know it's hard to precisely define, but in place like this an empty line would increase the readability of the file.

krzysztofrewak commented on backend/FileSystem/FileManager.php:
```diff
@@ -0,0 +1,39 @@
+class FileManager
+{
+    public function __construct(
+        protected DirectoryManager &$directoryManager
```
Why do you expecting explicite reference to the directory manager here?

plencka replied:
I was worried the object will be copied over, as I have not familiarized with PHP that much yet. Might be a bad habit from some issue I had in the past with something else.

krzysztofrewak commented on backend/FileSystem/FileManager.php:
```diff
@@ -0,0 +1,39 @@
+    ) {
+    }
+
+    public function create(string $relativePath, string $filename = "\0", mixed $content = ""): void
```
`\0`?

krzysztofrewak commented on backend/FileSystem/FileManager.php:
```diff
@@ -0,0 +1,39 @@
+        }
+    }
+
+    public function copyResource(string $relativeOrigin, string $relativeDestination, $filename, $newName = "\0"): void
```
```suggestion
    public function copyResource(string $relativeOrigin, string $relativeDestination, string $filename, string $newName = "\0"): void
```

krzysztofrewak commented on backend/FileSystem/FileManager.php:
```diff
@@ -0,0 +1,39 @@
+        $destination = $this->directoryManager->getApiPath($relativeDestination);
+        $origin = $this->directoryManager->getResourcePath($relativeOrigin);
+        if (!copy($origin . $filename, $destination . $newName)) {
+            throw new Exception("Couldn't copy. File " . $origin . $filename . " not found.");
```
I'm not frocing you to do this, but in PHP you can create your own exceptions to be more precise.

plencka replied:
I was considering custom exceptions when I research them, but I quickly threw it together to test feature I was working on first, then I kinda forgot about them. I'll toy around with this idea.

krzysztofrewak replied:
We can also create a new task to split exceptions. :)

plencka replied:
I think I'll open separate issue for that actually, it's not something that needs to be done ASAP, but would be nice to implement.

krzysztofrewak commented on backend/Models/Coordinates.php:
```diff
@@ -0,0 +1,45 @@
+    protected float $latitude;
+    protected float $longitude;
+
+    public function __construct(string $coordinates)
```
I would rather want to have a normal constructor with lat and lang. This parsing could be then moved to some other class.

krzysztofrewak commented on backend/Models/Coordinates.php:
```diff
@@ -0,0 +1,45 @@
+
+    protected function normalize(): void
+    {
+        $this->latitude = floatval(number_format($this->latitude, 6));
```
Are you sure that you want to use float casting here? http://sandbox.onlinephpfunctions.com/code/b191152d93c766b6c686c3833a713a6041abe017

plencka replied:
I see it cuts zeroes, but I'm not sure if that's an issue. Most applications don't display them and they aren't needed to be passed into MapBox.

Though I'm not sure if there's rounding issue, so I'll investigate that further.

krzysztofrewak commented on backend/Services/CsvReader.php:
```diff
@@ -0,0 +1,23 @@
+
+class CsvReader
+{
+    public function __construct()
```
Do you need this empty constructor here?

krzysztofrewak commented on backend/Services/DataBuilder.php:
```diff
@@ -0,0 +1,35 @@
+abstract class DataBuilder implements BuildTool
+{
+    protected string $workingDirectory;
+    protected static array $paths = [];
```
Oh, why would you need static field here?

plencka replied:
I want to create some manager to guard application from using same path + file combination for different data. I'm currently not sure how to create external class that will keep track of it from multiple builders, but current implementation is not something I want to keep.

And of course, I've just got possible idea how to do it while writing this comment, so I'll better implement it before it fades. External manager and passing paths into constructors should suffice. Path definitions will be moved ouside the Builders, but I think final result will increase readability.

krzysztofrewak reviewed (changes requested):
Nice work, but I expect some little changes and some explanations/motivations.

mlencki commented on backend/Services/CompanyDataBuilder.php:
```diff
@@ -0,0 +1,109 @@
+            true,
+            SANITIZE_WHITESPACE_REMOVE,
+            ",",
+            6
```
The `validateField` has quite a few arguments and it's hard to understand them all without looking into the method definition.

Maybe we can use named arguments here?
https://wiki.php.net/rfc/named_params

mlencki commented on backend/Services/DataBuilder.php:
```diff
@@ -0,0 +1,62 @@
+use Internships\Interfaces\SerializableInfo;
+use Internships\Models\PathPair;
+
+abstract class DataBuilder extends DataSanitizer implements BuildTool, SerializableInfo
```
Are you sure that `DataSanitizer` should be the parent of `DataBuilder`?

The name suggests different responsibilities of these classes.
I would use `DataSanitizer` as a dependency of `DataBuilder` using composition rather than inheritance.

mlencki commented on backend/Services/DataSanitizer.php:
```diff
@@ -0,0 +1,101 @@
+        return $value;
+    }
+
+    public function validateField(
```
Maybe we could narrow the responsibility of `DataSanitizer` by splitting this class into two? 
`DataSanitizer` and `DataValidator`?

The method signature suggests that we're validating a bunch of stuff here (dataName, id, requirement, some flags, decimals, etc.) but in fact, we're only checking the requirement of a given field, the rest of the arguments are for sanitizing data.

mtracz commented on backend/Application/Application.php:
```diff
@@ -0,0 +1,94 @@
+use Internships\Services\FacultyDataBuilder;
+use Internships\Services\UniquePathGuard;
+
+class Application
```
Maybe we should to think about dependency injection (DI)
https://github.com/PHP-DI/PHP-DI

plencka replied:
I'll investigate it through separate issue.

mtracz commented on backend/Application/Application.php:
```diff
@@ -0,0 +1,94 @@
+
+    public function build(): void
+    {
+        $jsonPrintFlags = JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT;
```
I thin we can add also `JSON_THROW_ON_ERROR | JSON_UNESCAPED_SLASHES`

mtracz commented on backend/FileSystem/DirectoryManager.php:
```diff
@@ -0,0 +1,45 @@
+    {
+        $directoryPath = $this->apiPath->getFull($relativePath);
+        if (!file_exists($directoryPath)) {
+            mkdir($directoryPath, 0777, true);
```
Let's move this raw data (0777, true) to some consts.

mtracz commented on backend/FileSystem/FileManager.php:
```diff
@@ -0,0 +1,42 @@
+    ) {
+    }
+
+    public function create(string $relativePath, string $filename = null, mixed $content = ""): void
```
Why `$filename` can be null?

plencka replied:
I intended this function to create both files or folders, even without files.

Maybe it could be used to create directory structure and then detect if any folder is missing some data.

mtracz replied:
Maybe add flag/param `$withoutFile = false` will be better?

mtracz commented on backend/FileSystem/FileManager.php:
```diff
@@ -0,0 +1,42 @@
+        string $relativeOrigin,
+        string $relativeDestination,
+        string $filename,
+        string $newName = ""
```
I think it should be null by default

mtracz commented on backend/Models/Address.php:
```diff
@@ -0,0 +1,61 @@
+
+    public function jsonSerialize(): array
+    {
+        return [
+            "coordinates" => $this->coordinates,
+            "countryLocaleCode" => $this->countryLocaleCode,
+            "cityLocaleCode" => $this->cityLocaleCode,
+            "street" => $this->street,
+            "zip" => $this->zip,
+        ];
```
```suggestion
        return get_object_vars($this);
```

krzysztofrewak replied:
I would stay with current mapping. It's easily configurable and more verbose.

mtracz replied:
Ok.

mtracz commented on backend/Models/Company.php:
```diff
@@ -0,0 +1,54 @@
+
+    public function jsonSerialize(): array
+    {
+        return [
+            "id" => $this->id,
+            "name" => $this->name,
+            "address" => $this->address,
+            "filterData" => $this->filterData,
+            "website" => $this->website,
+            "email" => $this->email,
+            "phoneNumber" => $this->phoneNumber,
+            "isPaid" => $this->isPaid,
+            "logoFile" => $this->logoFile,
+        ];
```
```suggestion
        return get_object_vars($this);
```

mtracz commented on backend/Models/Coordinates.php:
```diff
@@ -0,0 +1,37 @@
+
+    public function jsonSerialize(): array
+    {
+        return [
+            "latitude" => $this->latitude,
+            "longitude" => $this->longitude,
+        ];
```
```suggestion
        return get_object_vars($this);
```

mtracz commented on backend/Models/Faculty.php:
```diff
@@ -0,0 +1,45 @@
+
+    public function jsonSerialize(): array
+    {
+        return [
+            "id" => $this->id,
+            "name" => $this->name,
+            "directory" => $this->directory,
+        ];
```
```suggestion
        return get_object_vars($this);
```

mtracz commented on backend/Models/FilterData.php:
```diff
@@ -0,0 +1,37 @@
+
+    public function jsonSerialize(): array
+    {
+        return [
+            "specialization" => $this->specialization,
+            "tags" => $this->tags,
+        ];
```
```suggestion
        return get_object_vars($this);
```

mtracz commented on backend/Services/CompanyDataBuilder.php:
```diff
@@ -0,0 +1,109 @@
+    public function buildFromData(array $csvData): array
+    {
+        $companies = [];
+        foreach ($csvData as $i => $rowData) {
```
Let's rename `$i` to maybe `$rowNumber`?

mtracz commented on backend/Services/CompanyDataBuilder.php:
```diff
@@ -0,0 +1,109 @@
+                    "isPaid" => $rowData[11],
+                    "logoFile" => $rowData[12],
+                ];
+                array_push($companies, new Company($i - 1, $this->validate($i - 1, $entry)));
```
What is it `$i - 1`?

plencka replied:
I want entries to increment from 0, but the first entry is field definition.
It's skipped above at the beginning of foreach() statement.

Because javascript starts arrays from 0, I think it's better to handle it this way if we want to get filtered entries by ID without searching through them.

mtracz replied:
So let's move `$i - 1` to some variable, e.g. `$id`

mtracz commented on backend/Services/CsvReader.php:
```diff
@@ -0,0 +1,19 @@
+    public function getCSVData(string $fullPath): array
+    {
+        $csvRows = [];
+        $csvFile = fopen($fullPath, "r");
```
We can use here `try {} catch()` and throw exception to detect problems with file opening

mtracz commented on backend/Services/CsvReader.php:
```diff
@@ -0,0 +1,19 @@
+    {
+        $csvRows = [];
+        $csvFile = fopen($fullPath, "r");
+        while (($row = fgetcsv($csvFile, 0, ",")) !== false) {
```
Move this (`0, ","`) to the consts

mtracz commented on backend/Services/DataBuilder.php:
```diff
@@ -0,0 +1,62 @@
+    public function getSourceRelativePath(): string
+    {
+        return $this->source->getRelativePath()
+            . "/"
```
I would like to move `"/"` to the const, it's repeated very often

mtracz commented on backend/Services/CompanyDataBuilder.php:
```diff
@@ -0,0 +1,109 @@
+use Internships\Models\Company;
+use Internships\Models\PathPair;
+
+class CompanyDataBuilder extends DataBuilder implements SerializableInfo
```
It would be nice to test it out in unit tests.

krzysztofrewak replied:
There's #9 task for that.

plencka replied:
Tests are in separate issues.  They will be implemented when I'll learn how to make them.

mtracz commented on backend/FileSystem/DirectoryManager.php:
```diff
@@ -6,6 +6,9 @@
 
 class DirectoryManager
 {
+    public const DIRECTORY_PERMISSIONS = 0777;
+    public const RECURSIVE_CREATION = true;
```
Will these consts  be use somewhere else? If not, make them `protected`

mtracz commented on backend/FileSystem/DirectoryManager.php:
```diff
@@ -23,7 +26,7 @@ public function getApiDirectoryPath(string $relativePath): string
         $directoryPath = $this->apiPath->getFull($relativePath);
         if (!file_exists($directoryPath)) {
-            mkdir($directoryPath, 0777, true);
+            mkdir($directoryPath, self::DIRECTORY_PERMISSIONS, self::RECURSIVE_CREATION);
```
Try to use `static` instead `self`. It allow us to override constants in the child classes

mtracz commented on backend/Services/CsvReader.php:
```diff
@@ -4,16 +4,38 @@
+                throw new Exception("File " . $fullPath . " cannot be read");
+            }
+            while (($row = fgetcsv($csvFile, self::CSV_READ_LENGTH, self::CSV_SEPARATOR)) !== false) {
+                if ($rowCount = count($row)
```
I prefer not use assignments in the `if` statements. It is not readable

mtracz replied:
In this `while` loop there is assignment also.

mtracz commented on backend/Services/DataBuilder.php:
```diff
@@ -10,17 +10,43 @@
 
 abstract class DataBuilder extends DataSanitizer implements BuildTool, SerializableInfo
 {
+    private const FOLDER_SEPARATOR = "/";
```
Hmm. Maybe not define the same const, but use it from `Path` class?

mtracz commented on backend/FileSystem/Path.php:
```diff
@@ -4,26 +4,26 @@
 
 namespace Internships\FileSystem;
 
+define("PATHS_FOLDER_SEPARATOR", "/");
```
Hmm. Global consts are not the great idea. Let's back to the public const in the `Path` class.

plencka commented:
After I moved stuff around, I noted that it could be simplified even more.
Because of that, I introduced in the BuildTool interface the getModelClassToBuild() method. It returns the name of the Model class. It was a variable before that, but in my opinion method is more readable and forces developer to implement it.

I made it mainly to move buildFromData() function into base class, because only difference was Model used. Though I have no idea if that's correct way to do it, maybe there are better methods which I'm not aware about as of now.

mtracz commented on backend/FileSystem/Path.php:
```diff
@@ -4,26 +4,26 @@
     public function getFull($relativePath, bool $isFile = false): string
     {
-        $fullPath = $this->root . PATHS_FOLDER_SEPARATOR . $relativePath;
+        $fullPath = $this->root . self::FOLDER_SEPARATOR . $relativePath;
```
Why not `static`?

plencka replied:
Ah, I didn't consider this file much when making changes. Static makes more sense if somebody would like different path separator for various reasons.

mtracz commented on backend/Application/Application.php:
```diff
@@ -88,7 +92,32 @@ public function build(): void
+                $this->facultyBuilder->getSourceFileName()
+            ),
+            $this->facultyBuilder->getFields()
+        );
+        $templatePaths = $this->fileManager->getResourceFilePathsFrom($source);
+        foreach ($facultyCsvData as $rowNumber => $faculty) {
```
```suggestion
        );
        
        $templatePaths = $this->fileManager->getResourceFilePathsFrom($source);
        
        foreach ($facultyCsvData as $rowNumber => $faculty) {
```

mtracz commented on backend/Application/Application.php:
```diff
@@ -6,13 +6,15 @@
 
 class Application
 {
+    protected int $jsonPrintFlags;
```
```suggestion
    protected int $jsonPrintFlags = JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT | JSON_THROW_ON_ERROR | JSON_UNESCAPED_SLASHES;
```

mtracz commented on backend/Application/Application.php:
```diff
@@ -25,12 +27,14 @@ public function __construct(
         string $apiRelativePath,
         string $resourceRelativePath
     ) {
+        $this->jsonPrintFlags = JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT
+            | JSON_THROW_ON_ERROR | JSON_UNESCAPED_SLASHES;
```
```suggestion
```

mtracz commented on Dockerfile:
```diff
@@ -0,0 +1,15 @@
+FROM node:16-alpine3.14
```
Why this Dockerfile is in this MR?

plencka replied:
Based on merge history, it seems it suggested the other branch instead of the main I desired. I'm reverting that commit right now.
